### ISSUE TITLE
Enhanced Description of Passwordless SSH configurations

### ISSUE THREAD

**Issue by TheNotary** [bpg/terraform-provider-proxmox#619]
In reading the documentation at [terraform](https://registry.terraform.io/providers/bpg/proxmox/latest/docs#ssh-connection) it is states that:

> API Token authentication can be used to authenticate with the Proxmox API without the need to provide a password. In combination with the ssh block, this allows for a fully password-less authentication.

But it is unclear in what way the `ssh` block must be defined in order to achieve a password-less authentication.  On my local, if I run `ssh proxmox` I'm able to login to my proxmox server via public/ private key authentication.  Using many variations on the examples in the documentation has not produced successful results, rather the output I get from terraform is:

> Error: failed to dial IP_ADDRESS_TO_PROXMOX:22: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain

Does anyone know what I might be missing?  Here's where I left off with my provider block:

```
provider "proxmox" {
  endpoint = "https://example.com:8006"
  api_token = "${var.pm_api_token_id}=${var.pm_api_token_secret}"
  ssh {
    agent    = true
    # username = "root_plus"
    # address     = "proxmox"
    # node {
    #   name    = "proxmox"
    # }
  }
}
```

**Comment by otopetrik** [comment]
>  if I run `ssh proxmox` I'm able to login to my proxmox server via public/ private key authentication.

If agent is not running, `ssh` command itself can use ssh keys direclty from files on disk. It just asks for private key password before each use.

Terraform provider needs ssh-agent to be running.

1.  Verify that ssh-agent is running using `ssh-add -L`.

2. If it is not running (`Could not open a connection to your authentication agent.`), then start it:
```bash
eval $(ssh-agent)
```

3. If there are no keys loaded (`The agent has no identities.`), then load default one:
```bash
ssh-add
```
4. Try terraform in the *same* terminal.

Starting `ssh-agent` using the above method makes it avalable to the current terminal only.  Other terminal tabs/windows will not have `SSH_AUTH_SOCK` environment variable set. It may be good idea to configure operating system to auto-start `ssh-agent` early after login, making `SSH_AUTH_SOCK` available to all terminal windows.

**Comment by TheNotary** [comment]
Thanks for that info, sure enough ssh-agent didn't have a single key loaded until just now.

**Comment by TheNotary** [comment]
I've been thinking about this issue and am wondering if "enhancing the documentation" isn't the right direction.  I'm pretty unfamiliar with TF development, but I wonder if it's instead possible to throw a deliberate error/ warning from TF if the user specifies no password and `ssh-agent` has no keys loaded into it?  That way we save devs like me a trip to the docs .  

Btw, I think my laptop was in this state because I migrated to this laptop 3 years ago and didn't make any keys directly on it.  A multitude of tools I use rely on ssh, but apparently, not ssh-agent so I was really scratching my head for a while on this.  

If it would be helpful, I could set some time aside to add a PR for detecting this and printing an informative error.  I need a feature around creating qemu volumes (I'll open it when this issue closes) so will probably be digging around the code anyway.

**Comment by otopetrik** [comment]
> If it would be helpful, I could set some time aside to add a PR for detecting this and printing an informative error.

Detecting empty `ssh-agent` does not seem like the correct solution.
Just because `ssh-agent` does have *some* keys, it does not mean that it has the *correct* key.
e.g. an allways loaded "USB-token backed" `id_ecdsa_sk` for interactive use, and a "classic" key for use with Proxmox (to avoid need for repeated tapping of a button on USB-token), which is loaded by the user only as needed.

Looking at the code:
https://github.com/bpg/terraform-provider-proxmox/blob/aeb5e88bc9112686675c7058501fa9378b69af93/proxmox/ssh/client.go#L314-L336

It looks like the `failed to dial` error message appears only after successful connection to `ssh-agent` is verified, therefore we know there is ssh-agent running...

Maybe just create a PR modifying the error message.
Something like adding "(HINT: use 'ssh-add -L' to verify that ssh-agent has required key)" at the end of the message should be enough to help the user.

> I need a feature around creating qemu volumes (I'll open it when this issue closes) so will probably be digging around the code anyway.

Proxmox does not really like volume without and owners (= VM) - see [API docs](https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node}/storage/{storage}/content), POST call to allocate disk image requires vmid.

But it is possible for one VM to use a volume owned by another VM.
Instead of separate volume resource, a non-booting VM is used as a container for the volume.
Proof of concept of this functionality was implemented PR #606 and is already merged https://github.com/bpg/terraform-provider-proxmox/commit/aeb5e88bc9112686675c7058501fa9378b69af93.

**Comment by TheNotary** [comment]
I immensely appreciate that addition!  I can't wait to test it out.  Understood about it being a little tricky to pull off, I noticed that both Proxmox and qemu have an implementation of disks that work well for creating straightforward VMs, but doesn't have as much flexibility as other providers.  

I think you're right about wanting to just bring up `ssh-add -L` if the dial fails in that way.  I added a check for password field being blank in the PR.  Feel free to re-approach if it needs edits.  After struggling through the code-base and being new to golang I'm having trouble organizing my code, words, and even thoughts at this point #.#
